### PR TITLE
Enable partial bindless on Metal and reduce bind group overhead (#18149)

### LINKED ISSUE #18149: enable partial bindless support on metal
## What problem does this solve or what need does it fill?

Metal support some of the features for bindless, and could benefit from it

## What solution would you like?

> what I mean is that bindless is currently disabled unconditionally unless your platform supports bindless textures, bindless samplers, and bindless buffers. Metal supports the first two but not the last one. However, the last one is not necessary for StandardMaterial if you have flat buffers. So we don't actually need to disable bindless on Metal in most cases.

https://discord.com/channels/691052431525675048/743663924229963868/1346051575306780733

### PR DESCRIPTION
**Pull request by holg**
Make BUFFER_BINDING_ARRAY conditional on whether the material uses buffer binding arrays. Fix sampler limit to check array element count instead of binding slot count. Only create binding arrays for resource types the material actually uses, reducing overhead on all platforms.

## Objective
Fixes #18149
Metal supports `TEXTURE_BINDING_ARRAY` but not `BUFFER_BINDING_ARRAY`.
Bindless was disabled entirely on Metal because `bindless_supported()` required both unconditionally.

## Solution

1. **Conditional feature check**: Only require `BUFFER_BINDING_ARRAY` when the material actually uses buffer binding arrays. 
2. Materials using `#[data(...)]`, textures, and samplers (like `StandardMaterial`) only need `TEXTURE_BINDING_ARRAY`. **Fix sampler limit check**: Use `max_binding_array_sampler_elements_per_shader_stage` (array element count) instead of `max_samplers_per_shader_stage` (binding slot count).
3. **Only create needed binding arrays**: `create_bindless_bind_group_layout_entries` now skips resource types the material doesn't use. This stays within Metal's 31 argument buffer slot limit and reduces wasted fallback resources on all platforms.

## Testing
Bistro Exterior (698 materials), 5-minute runs:

  | GPU | Avg FPS (before → after) | Min FPS (before → after) | RAM/VRAM |
  |-----|--------------------------|--------------------------|----------|
  | Apple M2 Max (Metal) | 115 → 136 **(+18%)** | 60 → 106 **(+77%)** | -57 MB RAM |
  | NVIDIA 5060 Ti | 118 → 217 **(+84%)** | 60 → 165 **(+174%)** | Same |
  | Intel i360P | 25 → 29 **(+15%)** | Same | Same |
  | AMD Vega 8 / Ryzen 4800U | 25 → 25 | Same | **-88 MB VRAM** |
  | Intel Iris XE | ~22 → ~22 | Same | No regression |

Also tested: `3d_scene`, `pbr`, `lighting`, `transmission`, `deferred_rendering`: all pass with zero errors.
Materials using `#[uniform(..., binding_array(...))]` correctly fall back to non-bindless on Metal.

### CODE REVIEW

File: crates/bevy_render/src/render_resource/bindless.rs [hunk 1]
```diff
@@ -220,88 +220,85 @@ pub struct BindlessIndex(pub u32);
...
-/// `bindless_slab_resource_limit` specifies the resolved
-/// [`BindlessSlabResourceLimit`] value.
+/// `used_resource_types` limits which binding arrays are created,
+/// reducing argument buffer slot usage on constrained platforms.
 pub fn create_bindless_bind_group_layout_entries(
     bindless_index_table_length: u32,
     bindless_slab_resource_limit: u32,
     bindless_index_table_binding_number: BindingNumber,
+    used_resource_types: &[BindlessResourceType],
 ) -> Vec<BindGroupLayoutEntry> {
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_render/src/render_resource/bindless.rs:230]: -1, we don't enforce API stability in rendering guts

File: crates/bevy_render/src/render_resource/bindless.rs [hunk 2]
```diff
@@ -220,88 +220,85 @@ pub struct BindlessIndex(pub u32);
...
-/// `bindless_resource_count` specifies the total number of bindless resources.
-/// `bindless_slab_resource_limit` specifies the resolved
-/// [`BindlessSlabResourceLimit`] value.
+/// `used_resource_types` limits which binding arrays are created,
+/// reducing argument buffer slot usage on constrained platforms.
 pub fn create_bindless_bind_group_layout_entries(
     bindless_index_table_length: u32,
     bindless_slab_resource_limit: u32,
     bindless_index_table_binding_number: BindingNumber,
+    used_resource_types: &[BindlessResourceType],
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_render/src/render_resource/bindless.rs:229]: -1, the comments were out of date

File: crates/bevy_render/macros/src/as_bind_group.rs [hunk 3]
```diff
@@ -73,6 +73,9 @@ pub fn derive_as_bind_group(ast: syn::DeriveInput) -> Result<TokenStream> {
     let mut non_bindless_binding_layouts = Vec::new();
     let mut bindless_resource_types = Vec::new();
     let mut bindless_buffer_descriptors = Vec::new();
+    // Whether this material needs buffer binding arrays (BUFFER_BINDING_ARRAY feature).
+    // False when only #[data(...)], textures, and samplers are used.
```
Comment (pcwalton) [inline crates/bevy_render/macros/src/as_bind_group.rs:77]: I might mention in the comment for future readers that if and when wgpu adds support for buffer binding arrays in Metal, we can remove this.

Comment (mate-h) [reply crates/bevy_render/macros/src/as_bind_group.rs:77]: my PR is under review for adding support for buffer binding arrays in Metal: https://github.com/gfx-rs/wgpu/pull/9081

**Comment by kfc35** [comment]
> It looks like your PR has been selected for a highlight in the next release blog post, but **you didn't provide a release note**.
> 
> Please review the [instructions for writing release notes](https://github.com/bevyengine/bevy/tree/main/release-content/release_notes.md), then expand or revise the content in the [release notes directory](https://github.com/bevyengine/bevy/tree/main/release-content/release-notes) to showcase your changes.

This bot is pointing to the wrong links (this has already been fixed)
Here are the corrected links:
directions: https://github.com/bevyengine/bevy/blob/main/_release-content/release_notes.md

directory: https://github.com/bevyengine/bevy/tree/main/_release-content/release-notes

Once you’ve made one, I’ll do a quick look over and place this into ready for final review 😄

**Comment by holg** [comment]
Added release note in _release-content/release-notes/partial_bindless_metal.md. Also addressed pcwalton's nit about the future Metal buffer binding array support comment (referencing wgpu#9081) in the previous commit.

File: crates/bevy_render/macros/src/as_bind_group.rs [hunk 4]
```diff
@@ -1090,12 +1112,16 @@ pub fn derive_as_bind_group(ast: syn::DeriveInput) -> Result<TokenStream> {
                 match #actual_bindless_slot_count {
                     Some(bindless_slot_count) => {
                         let bindless_index_table_range = #bindless_index_table_range;
+                        let used_resource_types: &[#render_path::render_resource::BindlessResourceType] = &[
```
Comment (mate-h) [inline crates/bevy_render/macros/src/as_bind_group.rs]: Code style nit pick: does the type need to be explicitly declared here or can it be inferred? Also if it does need to be declared then shorten it by adding an import for that type.

File: crates/bevy_render/src/render_resource/bindless.rs [hunk 5]
```diff
@@ -220,88 +220,85 @@ pub struct BindlessIndex(pub u32);
...
-            .build(
-                9,
-                ShaderStages::FRAGMENT | ShaderStages::VERTEX | ShaderStages::COMPUTE,
-            ),
-    ]
+        .build(*bindless_index_table_binding_number, stages),
+    ];
+
+    // binding arrays only for types that this material uses
+    // Specifically Metal where each binding array uses the buffer slot (limited to 31)
```
Comment (mate-h) [inline crates/bevy_render/src/render_resource/bindless.rs]: Specify where this limit comes from ie the metal feature set tables or reference a wgpu constant

File: crates/bevy_render/src/render_resource/bindless.rs [hunk 6]
```diff
@@ -220,88 +220,85 @@ pub struct BindlessIndex(pub u32);
...
+    for &(resource_type, ref binding_number) in BINDING_NUMBERS.iter() {
+        if !used_resource_types.contains(&resource_type) {
+            continue;
+        }
+        let entry = match resource_type {
+            BindlessResourceType::SamplerFiltering => sampler(SamplerBindingType::Filtering)
+                .count(bindless_slab_resource_limit)
+                .build(**binding_number, stages),
```
Comment (mate-h) [inline crates/bevy_render/src/render_resource/bindless.rs]: These .count and .build lines are repeated frequently. Maybe reorganize the code so that first you return Some type from the match and then count and build. For the ones with a continue map to None 
And skip none after for counting

File: crates/bevy_render/macros/src/as_bind_group.rs (hunk 4 above)
Comment (holg) [reply crates/bevy_render/macros/src/as_bind_group.rs]: Fair point, it can be inferred from the context..., will change...

File: _release-content/release-notes/partial_bindless_metal.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+---
```
Comment (kfc35) [inline _release-content/release-notes/partial_bindless_metal.md:1]: Hm, this file still ended up in `release-content` and not `_release-content` 😅 maybe `main` needs to be merged into your branch if the directory hasn’t been updated

**Comment by kfc35** [comment]
I’ve put this in final review but a note for the maintainer that the release note needs to be moved!
